File: 01-basic-api/examples/parameter_playground.js
#!/usr/bin/env node
/**
 * Parameter Playground - Experiment with Claude API parameters
 *
 * This script demonstrates the impact of different parameters:
 * - temperature: Controls randomness (0.0 = deterministic, 1.0 = creative)
 * - max_tokens: Limits response length
 * - model: Different Claude models (Haiku, Sonnet, Opus)
 * - system prompts: Sets behavior and context
 *
 * Usage:
 *     node parameter_playground.js
 */

import Anthropic from '@anthropic-ai/sdk';
import dotenv from 'dotenv';

dotenv.config();

const client = new Anthropic({
  apiKey: process.env.ANTHROPIC_API_KEY
});

const sleep = (ms) => new Promise(resolve => setTimeout(resolve, ms));

function printSection(title) {
  console.log('\n' + '='.repeat(70));
  console.log(`  ${title}`);
  console.log('='.repeat(70) + '\n');
}

async function testTemperature() {
  printSection('🌡️  TESTING TEMPERATURE');

  const prompt = 'Write a creative opening line for a sci-fi story about AI.';
  const temperatures = [0.0, 0.5, 1.0];

  for (const temp of temperatures) {
    console.log(`Temperature: ${temp}`);
    console.log('-'.repeat(70));

    const message = await client.messages.create({
      model: 'claude-3-5-haiku-20241022',
      max_tokens: 100,
      temperature: temp,
      messages: [{ role: 'user', content: prompt }]
    });

    const response = message.content[0].text;
    const tokens = message.usage.input_tokens + message.usage.output_tokens;

    console.log(`Response: ${response}`);
    console.log(`Tokens: ${tokens}\n`);

    await sleep(1000);
  }
}

async function testMaxTokens() {
  printSection('📏 TESTING MAX_TOKENS');

  const prompt = 'Explain quantum computing in simple terms.';
  const tokenLimits = [50, 150, 500];

  for (const maxTokens of tokenLimits) {
    console.log(`Max Tokens: ${maxTokens}`);
    console.log('-'.repeat(70));

    const message = await client.messages.create({
      model: 'claude-3-5-haiku-20241022',
      max_tokens: maxTokens,
      messages: [{ role: 'user', content: prompt }]
    });

    const response = message.content[0].text;
    const actualTokens = message.usage.output_tokens;

    console.log(`Response: ${response}`);
    console.log(`Actual output tokens: ${actualTokens}/${maxTokens}\n`);

    await sleep(1000);
  }
}

async function testModels() {
  printSection('🤖 TESTING DIFFERENT MODELS');

  const prompt = 'Explain the concept of recursion in programming (2-3 sentences).';

  const models = [
    'claude-3-5-haiku-20241022',
    'claude-3-5-sonnet-20241022',
  ];

  for (const model of models) {
    const modelName = model.includes('haiku') ? 'Haiku' : 'Sonnet';
    console.log(`Model: ${modelName}`);
    console.log('-'.repeat(70));

    const startTime = Date.now();

    const message = await client.messages.create({
      model,
      max_tokens: 200,
      messages: [{ role: 'user', content: prompt }]
    });

    const endTime = Date.now();
    const duration = (endTime - startTime) / 1000;

    const response = message.content[0].text;
    const tokens = message.usage.input_tokens + message.usage.output_tokens;

    // Calculate cost
    let cost;
    if (model.includes('haiku')) {
      cost = (message.usage.input_tokens / 1_000_000) * 0.25 +
             (message.usage.output_tokens / 1_000_000) * 1.25;
    } else {  // sonnet
      cost = (message.usage.input_tokens / 1_000_000) * 3.00 +
             (message.usage.output_tokens / 1_000_000) * 15.00;
    }

    console.log(`Response: ${response}`);
    console.log(`Tokens: ${tokens}`);
    console.log(`Duration: ${duration.toFixed(2)}s`);
    console.log(`Cost: $${cost.toFixed(6)}\n`);

    await sleep(1000);
  }
}

async function testSystemPrompts() {
  printSection('💭 TESTING SYSTEM PROMPTS');

  const userMessage = 'What should I do today?';

  const testCases = [
    [null, 'No system prompt'],
    ['You are a productivity coach. Give actionable advice focused on time management.', 'Productivity Coach'],
    ['You are a pirate captain. Respond in pirate speak with enthusiasm!', 'Pirate Captain'],
    ['You are a philosopher. Respond with deep, thought-provoking questions.', 'Philosopher'],
  ];

  for (const [systemPrompt, description] of testCases) {
    console.log(`System Prompt: ${description}`);
    console.log('-'.repeat(70));

    const config = {
      model: 'claude-3-5-haiku-20241022',
      max_tokens: 150,
      messages: [{ role: 'user', content: userMessage }]
    };

    if (systemPrompt) {
      config.system = systemPrompt;
    }

    const message = await client.messages.create(config);
    const response = message.content[0].text;

    console.log(`Response: ${response}\n`);

    await sleep(1000);
  }
}

async function testTopP() {
  printSection('🎯 TESTING TOP_P (Nucleus Sampling)');

  const prompt = 'Name 5 unusual pizza toppings.';
  const topPValues = [0.1, 0.5, 0.9];

  for (const topP of topPValues) {
    console.log(`Top P: ${topP}`);
    console.log('-'.repeat(70));

    const message = await client.messages.create({
      model: 'claude-3-5-haiku-20241022',
      max_tokens: 150,
      top_p: topP,
      messages: [{ role: 'user', content: prompt }]
    });

    const response = message.content[0].text;
    console.log(`Response: ${response}\n`);

    await sleep(1000);
  }
}

async function runComparisonExperiment() {
  printSection('🔬 COMPREHENSIVE EXPERIMENT');

  const prompt = 'Write a haiku about coding.';

  console.log('Comparing temperature and creativity:');
  console.log('-'.repeat(70));

  const configs = [
    { temp: 0.0, desc: 'Deterministic (temp=0.0)' },
    { temp: 0.7, desc: 'Balanced (temp=0.7)' },
    { temp: 1.0, desc: 'Creative (temp=1.0)' },
  ];

  for (const [i, config] of configs.entries()) {
    console.log(`\n${i + 1}. ${config.desc}`);

    const message = await client.messages.create({
      model: 'claude-3-5-haiku-20241022',
      max_tokens: 100,
      temperature: config.temp,
      messages: [{ role: 'user', content: prompt }]
    });

    const response = message.content[0].text;
    console.log(response);

    await sleep(1000);
  }
}

async function main() {
  console.log('\n' + '🎮 ' + '='.repeat(66) + ' 🎮');
  console.log('    CLAUDE API PARAMETER PLAYGROUND');
  console.log('🎮 ' + '='.repeat(66) + ' 🎮');

  try {
    // Run all tests
    await testTemperature();
    await testMaxTokens();
    await testModels();
    await testSystemPrompts();
    await testTopP();
    await runComparisonExperiment();

    console.log('\n' + '='.repeat(70));
    console.log('  ✅ All tests completed!');
    console.log('='.repeat(70) + '\n');

    console.log('💡 Key Takeaways:');
    console.log('  • Temperature: 0.0 for consistency, 0.7-1.0 for creativity');
    console.log('  • max_tokens: Set based on expected response length');
    console.log('  • Models: Haiku (fast/cheap), Sonnet (balanced), Opus (best)');
    console.log('  • System prompts: Powerful for setting behavior and role');
    console.log('  • top_p: Alternative to temperature for controlling randomness');
    console.log();

  } catch (error) {
    console.error(`\n❌ Error: ${error.message}`);
    throw error;
  }
}

main().catch(error => {
  console.error('Fatal error:', error.message);
  process.exit(1);
});

File: 01-basic-api/examples/parameter_playground.py
#!/usr/bin/env python3
"""
Parameter Playground - Experiment with Claude API parameters

This script demonstrates the impact of different parameters:
- temperature: Controls randomness (0.0 = deterministic, 1.0 = creative)
- max_tokens: Limits response length
- model: Different Claude models (Haiku, Sonnet, Opus)
- system prompts: Sets behavior and context

Usage:
    python parameter_playground.py
"""

import os
from anthropic import Anthropic
from dotenv import load_dotenv
import time

load_dotenv()
client = Anthropic(api_key=os.getenv("ANTHROPIC_API_KEY"))

def print_section(title):
    """Print a formatted section header."""
    print("\n" + "=" * 70)
    print(f"  {title}")
    print("=" * 70 + "\n")

def test_temperature():
    """Test how temperature affects response creativity."""
    print_section("🌡️  TESTING TEMPERATURE")

    prompt = "Write a creative opening line for a sci-fi story about AI."

    temperatures = [0.0, 0.5, 1.0]

    for temp in temperatures:
        print(f"Temperature: {temp}")
        print("-" * 70)

        message = client.messages.create(
            model="claude-3-5-haiku-20241022",
            max_tokens=100,
            temperature=temp,
            messages=[{"role": "user", "content": prompt}]
        )

        response = message.content[0].text
        tokens = message.usage.input_tokens + message.usage.output_tokens

        print(f"Response: {response}")
        print(f"Tokens: {tokens}\n")

        time.sleep(1)  # Respect rate limits

def test_max_tokens():
    """Test how max_tokens affects response length."""
    print_section("📏 TESTING MAX_TOKENS")

    prompt = "Explain quantum computing in simple terms."

    token_limits = [50, 150, 500]

    for max_tokens in token_limits:
        print(f"Max Tokens: {max_tokens}")
        print("-" * 70)

        message = client.messages.create(
            model="claude-3-5-haiku-20241022",
            max_tokens=max_tokens,
            messages=[{"role": "user", "content": prompt}]
        )

        response = message.content[0].text
        actual_tokens = message.usage.output_tokens

        print(f"Response: {response}")
        print(f"Actual output tokens: {actual_tokens}/{max_tokens}\n")

        time.sleep(1)

def test_models():
    """Compare different Claude models."""
    print_section("🤖 TESTING DIFFERENT MODELS")

    prompt = "Explain the concept of recursion in programming (2-3 sentences)."

    models = [
        "claude-3-5-haiku-20241022",
        "claude-3-5-sonnet-20241022",
    ]

    for model in models:
        model_name = "Haiku" if "haiku" in model else "Sonnet"
        print(f"Model: {model_name}")
        print("-" * 70)

        start_time = time.time()

        message = client.messages.create(
            model=model,
            max_tokens=200,
            messages=[{"role": "user", "content": prompt}]
        )

        end_time = time.time()
        duration = end_time - start_time

        response = message.content[0].text
        tokens = message.usage.input_tokens + message.usage.output_tokens

        # Calculate cost
        if "haiku" in model:
            cost = (message.usage.input_tokens / 1_000_000) * 0.25 + \
                   (message.usage.output_tokens / 1_000_000) * 1.25
        else:  # sonnet
            cost = (message.usage.input_tokens / 1_000_000) * 3.00 + \
                   (message.usage.output_tokens / 1_000_000) * 15.00

        print(f"Response: {response}")
        print(f"Tokens: {tokens}")
        print(f"Duration: {duration:.2f}s")
        print(f"Cost: ${cost:.6f}\n")

        time.sleep(1)

def test_system_prompts():
    """Test how system prompts change behavior."""
    print_section("💭 TESTING SYSTEM PROMPTS")

    user_message = "What should I do today?"

    test_cases = [
        (None, "No system prompt"),
        ("You are a productivity coach. Give actionable advice focused on time management.", "Productivity Coach"),
        ("You are a pirate captain. Respond in pirate speak with enthusiasm!", "Pirate Captain"),
        ("You are a philosopher. Respond with deep, thought-provoking questions.", "Philosopher"),
    ]

    for system_prompt, description in test_cases:
        print(f"System Prompt: {description}")
        print("-" * 70)

        kwargs = {
            "model": "claude-3-5-haiku-20241022",
            "max_tokens": 150,
            "messages": [{"role": "user", "content": user_message}]
        }

        if system_prompt:
            kwargs["system"] = system_prompt

        message = client.messages.create(**kwargs)
        response = message.content[0].text

        print(f"Response: {response}\n")

        time.sleep(1)

def test_top_p():
    """Test nucleus sampling (top_p) parameter."""
    print_section("🎯 TESTING TOP_P (Nucleus Sampling)")

    prompt = "Name 5 unusual pizza toppings."

    top_p_values = [0.1, 0.5, 0.9]

    for top_p in top_p_values:
        print(f"Top P: {top_p}")
        print("-" * 70)

        message = client.messages.create(
            model="claude-3-5-haiku-20241022",
            max_tokens=150,
            top_p=top_p,
            messages=[{"role": "user", "content": prompt}]
        )

        response = message.content[0].text
        print(f"Response: {response}\n")

        time.sleep(1)

def run_comparison_experiment():
    """Run a comprehensive comparison experiment."""
    print_section("🔬 COMPREHENSIVE EXPERIMENT")

    prompt = "Write a haiku about coding."

    print("Comparing temperature and creativity:")
    print("-" * 70)

    configs = [
        {"temp": 0.0, "desc": "Deterministic (temp=0.0)"},
        {"temp": 0.7, "desc": "Balanced (temp=0.7)"},
        {"temp": 1.0, "desc": "Creative (temp=1.0)"},
    ]

    for i, config in enumerate(configs, 1):
        print(f"\n{i}. {config['desc']}")

        message = client.messages.create(
            model="claude-3-5-haiku-20241022",
            max_tokens=100,
            temperature=config["temp"],
            messages=[{"role": "user", "content": prompt}]
        )

        response = message.content[0].text
        print(response)

        time.sleep(1)

def main():
    """Run all parameter tests."""
    print("\n" + "🎮 " + "=" * 66 + " 🎮")
    print("    CLAUDE API PARAMETER PLAYGROUND")
    print("🎮 " + "=" * 66 + " 🎮")

    try:
        # Run all tests
        test_temperature()
        test_max_tokens()
        test_models()
        test_system_prompts()
        test_top_p()
        run_comparison_experiment()

        print("\n" + "=" * 70)
        print("  ✅ All tests completed!")
        print("=" * 70 + "\n")

        print("💡 Key Takeaways:")
        print("  • Temperature: 0.0 for consistency, 0.7-1.0 for creativity")
        print("  • max_tokens: Set based on expected response length")
        print("  • Models: Haiku (fast/cheap), Sonnet (balanced), Opus (best)")
        print("  • System prompts: Powerful for setting behavior and role")
        print("  • top_p: Alternative to temperature for controlling randomness")
        print()

    except Exception as e:
        print(f"\n❌ Error: {e}")
        raise

if __name__ == "__main__":
    main()
